### PR TITLE
Make MountAPIVFS=yes default

### PR DESCRIPTION
**Pull request by keszybz**
Another chunk from #17079.

### CODE REVIEW

File: src/core/load-fragment.c [hunk 1]
```diff
@@ -1349,6 +1349,44 @@ int config_parse_exec_cpu_sched_policy(const char *unit,
         return 0;
 }
 
+int config_parse_exec_mount_apivfs(const char *unit,
```
Comment (poettering) [inline src/core/load-fragment.c:1352]: we usually line break function headers after the `(` these days

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
hmm, so this looks good codewise and everything

but i wonder: wouldn't it actually be sufficient to just the flip the default? iirc MountAPIVFS= is a NOP anyway if /proc is already mounted, no? And if that's the case the fact that with this PR MountAPIVFS= remains off by default unless RootImage=/RootDirectory= are set doesn't get us anything, since even if it was on then it wouldn't be doing anything anyway.

or am I confused?

**Comment by keszybz** [comment]
> but i wonder: wouldn't it actually be sufficient to just the flip the default? iirc MountAPIVFS= is a NOP anyway if /proc is already mounted, no? And if that's the case the fact that with this PR MountAPIVFS= remains off by default unless RootImage=/RootDirectory= are set doesn't get us anything, since even if it was on then it wouldn't be doing anything anyway.

I wanted to keep the possibility to have RootDirectory= with static /proc,/dev. I was thinking about the case how people set up chroots for programs traditionally: create a subdirectory with the binaries and the minimum data files and the minimum device nodes required. The sad truth is that even with hidepid=, /proc exposes waaaay too much information. We'd still support such setups except that people would need to specify MountAPIVFS=false.

**Comment by poettering** [comment]
But that would still work: just set MountAPIVFS=no, just like in your case.

My suggestion would be: just flip the default to be MountAPIVFS=yes. this would have zero effect if /proc/ and /sys/ are already mounted. And if they aren#t mounted it would mean we'd start mounting them now. But if the user sets =no explicitly we'd not do that either.

See the path_is_mount_point() checks in namespace.c's mount_sysfs(), mount_procfs() and mount_bind_dev(): they have no effect if there's already a mount point in place, which means we can just turn MountAPIVFS= on by default, and thus make it opt-out instead of opt-in.

**Comment by keszybz** [comment]
Indeed, I don't know why I was trying to make it so complicated.

**Comment by keszybz** [comment]
One centosci failed in TEST-43-PRIVATEUSER-UNPRIV and the other in TEST-44-LOG-NAMESPACE, probably unrelated.

File: src/core/mount.c [hunk 2]
```diff
@@ -160,6 +160,8 @@ static void mount_init(Unit *u) {
          * already trying to comply its last one. */
         m->exec_context.same_pgrp = true;
 
+        m->exec_context.mount_apivfs = true;
```
Comment (poettering) [inline src/core/mount.c]: hmm, actually, this doesn't look right... right now if the bool is set this has the effect that a mount namespace is created... (see exec_needs_mount_namespace()) which we actually don#t want here. (in particular not for mount units...) my bad, sorry for the confusion

I figure we have to have this as two bools after all, my bad. When the bool is unset turn it on but do not enable the namespace logic, when the bool is true enable the namespace logic and turn it off, and when is off do not propagate that to the namespace logic.

**Comment by poettering** [comment]
I suck, my bad, previous version of the patch was actually more correct I guess. Sorry!

**Comment by keszybz** [comment]
Reverted to the previous version.

File: src/core/execute.c [hunk 3]
```diff
@@ -5653,9 +5653,14 @@ int exec_context_get_effective_ioprio(const ExecContext *c) {
 bool exec_context_get_effective_mount_apivfs(const ExecContext *c) {
         assert(c);
 
+        /* Explicit setting wins */
         if (c->mount_apivfs_set)
                 return c->mount_apivfs;
 
+        /* Default to "yes" if root directory or image are specified */
+        if (c->root_image || !empty_or_root(c->root_directory))
```
Comment (poettering) [inline src/core/execute.c:5661]: i added #17219 now which should make this a bit prettier

**Pull request merged**